### DISCUSSION TITLE
Accept null as false-y value in conditional expressions

### DISCUSSION THREAD

**Discussion by michael-hawker** [dotnet/csharplang#4707] [Language Ideas]
The Null-conditional operator has been a great add to the language, and I love writing "obj?.prop" over having to check for null first in a lot of cases. However, this starts to break down in some if statements.

For instance the following scenario:

```cs
if (dict != null && dict.TryGetValue(key, out object value))
```

It'd be great if the `if` statement could accept `null` as an alternative to `false` as well, then I could write this:

```cs
if (dict?.TryGetValue(key, out object value))
```

Not sure if this should generate a warning, but the null-forgiving operator could be used at the end to suppress that if an extra-validation is thought to be needed to say that the programmer wants to accept this behavior.

Thoughts?

**Comment by HaloFour** [comment]
But [`null`](https://ericlippert.com/2012/03/26/null-is-not-false-part-one/) [is not](https://ericlippert.com/2012/04/12/null-is-not-false-part-two/) [`false`](https://ericlippert.com/2012/04/19/null-is-not-false-part-three/).

**Reply by michael-hawker** [reply]
Yeah, I know that, but in this case I don't care that it's null. I only care for the if to evaluate if it's `true`, and the following doesn't work either with how the short-cutting works:

```cs
if (dict?.TryGetValue(key, out object value) is true)
```

You'll get a `value` may not be defined error.

**Reply by HaloFour** [reply]
IMO that would be more of an ask to rework definite assignment when it comes to null-propagation, `out` declarations and checking the return type.  As you can see by using the pattern `true or not null` the behavior is the same, so treating `null` as `false` wouldn't have an effect on that error.

**Reply by michael-hawker** [reply]
Sure, I think there's a couple of things going on here. Ultimately, I just want to simplify the statement here. 🙂

**Reply by 333fred** [reply]
The definite assignment bit will be fixed in C# 10: https://github.com/dotnet/csharplang/issues/4465.

**Reply by HaloFour** [reply]
> Ultimately, I just want to simplify the statement here.

I think it might look simpler but would ultimately be a lot more complicated.  There is simplicity in the compiler only considering `bool` expressions to be `true`/`false`.  If you start to wade into "truthiness" it opens cans of worms when it comes to interpreting the intent of the developer.  I think the three articles I linked illustrate this problem quite well.

**Comment by 333fred** [comment]
See also:

https://github.com/dotnet/csharplang/discussions/2861
https://github.com/dotnet/csharplang/issues/438
https://github.com/dotnet/csharplang/discussions/134

**Reply by michael-hawker** [reply]
Thanks Fred! I also stumbled upon this one: https://github.com/dotnet/csharplang/issues/1410
